ocaml/testsuite/tests/typing-unique/overwriting.ml

```diff
@@ -661,6 +787,169 @@ Error: Overwrite may not change the tag to OptionA.
+4 |     (r.m <- OptionB s), overwrite_ r.m with OptionA s
+                                                ^^^^^^^
+Error: Overwrite may not change the tag to OptionA.
+Hint: The old tag of this allocation was changed through mutation.
+|}]
+
+let mutating_tag_par_parent r =
+  match r with
+  | { m = { x = OptionA s } } ->
+    (r.m <- { x = OptionB s }), overwrite_ r.m.x with OptionA s
+  | _ -> (), OptionB ""
+[%%expect{|
+Line 4, characters 54-61:
+4 |     (r.m <- { x = OptionB s }), overwrite_ r.m.x with OptionA s
+                                                          ^^^^^^^
+Error: Overwrite may not change the tag to OptionA.
+Hint: The old tag of this allocation was changed through mutation.
+|}]
+
+let mutating_tag_choice r =
+  match r with
+  | { m = OptionA s } ->
+    if true then (r.m <- OptionB s; r.m)
+            else overwrite_ r.m with OptionA s
+  | _ -> OptionB ""
+[%%expect{|
+Line 5, characters 37-44:
+5 |             else overwrite_ r.m with OptionA s
+                                         ^^^^^^^
+Error: Overwrite may not change the tag to OptionA.
+Hint: The old tag of this allocation was changed through mutation.
+|}]
+
+let mutating_tag_choice_parent r =
+  match r with
+  | { m = { x = OptionA s } } ->
+    if true then (r.m <- { x = OptionB s }; r.m.x)
+    else overwrite_ r.m.x with OptionA s
+  | _ -> OptionB ""
+[%%expect{|
+Line 5, characters 31-38:
+5 |     else overwrite_ r.m.x with OptionA s
+                                   ^^^^^^^
+Error: Overwrite may not change the tag to OptionA.
+Hint: The old tag of this allocation was changed through mutation.
+|}]
+
+let mutating_tag_choice_seq r =
+  match r with
+  | { m = OptionA s } ->
+    (if true then r.m <- OptionB s else ());
+    overwrite_ r.m with OptionA s
+  | _ -> OptionB ""
+[%%expect{|
+Line 5, characters 24-31:
+5 |     overwrite_ r.m with OptionA s
+                            ^^^^^^^
+Error: Overwrite may not change the tag to OptionA.
+Hint: The old tag of this allocation was changed through mutation.
+|}]
+
+let mutating_tag_choice_seq_parent r =
+  match r with
+  | { m = { x = OptionA s } } ->
+    (if true then r.m <- { x = OptionB s } else ());
+    overwrite_ r.m.x with OptionA s
+  | _ -> OptionB ""
+[%%expect{|
+Line 5, characters 26-33:
+5 |     overwrite_ r.m.x with OptionA s
+                              ^^^^^^^
+Error: Overwrite may not change the tag to OptionA.
+Hint: The old tag of this allocation was changed through mutation.
+|}]
+
+
+let mutating_tag_rematch r =
+  match r with
+  | { m = OptionA s } ->
+    r.m <- OptionB s;
+    begin match r with
+    | { m = OptionB s } ->
+      overwrite_ r.m with OptionB s
+    | _ -> OptionB ""
+    end
+  | _ -> OptionB ""
+[%%expect{|
+Line 7, characters 26-33:
+7 |       overwrite_ r.m with OptionB s
+                              ^^^^^^^
+Error: Overwrite may not change the tag to OptionB.
+Hint: The old tag of this allocation was changed through mutation.
+|}]
+
```

Comment:
What about something where the mutation doesn't change the tag? Is that allowed?

Reply:
No, all mutation is bad. I added a test case showing that we don't try to determine the tag statically